### PR TITLE
[PyROOT] Provide an explanatory error message when pickling cross-inh…

### PR DESCRIPTION
**Pull request by etejedor**
…eritance Python objects

When a Python class derives from a C++ class, cppyy jits a C++ wrapper class
that inherits from the base C++	class. Since I/O on jitted classes is not
supported by ROOT, trying to pickle a cross-inheritance Python object fails.
For this to work, dictionaries would need to be generated for the jitted class
before serialization, and they would need to be carried along for deserialization too.

Therefore, this commit replaces the previous segfault by an explanatory error
message when the user tries to pickle a cross-inheritance Python object. It is
suggested in the message to define a custom `__reduce__` method for the derived
Python class. That `__reduce__` method can e.g. return a callable and some
arguments with which the derived object can be constructed during deserialization,
thus preventing any attempt to serialize the object itself.

The old PyROOT did not jit any wrapper class during cross-inheritance (so it did
not suffer from the problem described above), but serialization/deserialization
of cross-inheritance Python objects was not properly supported anyway,
since the class of the unpickled object was the base C++ class, not the derived
Python class.

### CODE REVIEW

File: bindings/pyroot/pythonizations/src/CPPInstancePyz.cxx [hunk 1]
```diff
@@ -83,13 +83,19 @@ PyObject *op_reduce(CPPInstance *self, PyObject * /*args*/)
    if (s_bfClass == self->ObjectIsA()) {
       buff = (TBufferFile *)self->GetObject();
    } else {
+      auto className = Cppyy::GetScopedFinalName(self->ObjectIsA());
+      if (className.find("__cppyy_internal::Dispatcher") == 0) {
+         PyErr_Format(PyExc_IOError, "cannot stream Python objects whose class derives from a C++ class. "
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/src/CPPInstancePyz.cxx:88]: Maybe specify that it cannot be streamed *by default* and that the custom `__reduce__` function can make the serialization work. Also maybe something like "Refer to the python pickle package standard documentation for help".

**Review by vepadulano — APPROVED** [review]
It's good that we avoid the segfault! I would maybe spend a few words more in the error message, up to you

**Comment by vepadulano** [comment]
Now that I think about it, it would be good to have a test for this using maybe the reproducer provided at https://root-forum.cern.ch/t/new-pyroot-cant-pickle-derived-class-types/45415

File: bindings/pyroot/pythonizations/src/CPPInstancePyz.cxx (hunk 1 above)
Comment (etejedor) [reply bindings/pyroot/pythonizations/src/CPPInstancePyz.cxx]: Done!

**Comment by etejedor** [comment]
Sure I'll add the test too!

**Comment by etejedor** [comment]
Test: https://github.com/root-project/roottest/pull/722

**Pull request merged**